marimo/_plugins/ui/_impl/dataframes/transforms/apply.py

```diff
@@ -0,0 +1,142 @@
+        assert_never(transform.type)
+
+
+def apply_transforms(
+    df: T, handler: TransformHandler[T], transforms: Transformations
+) -> T:
+    if not transforms.transforms:
+        return df
+    for transform in transforms.transforms:
+        df = handle(df, handler, transform)
+    return df
+
+
+def get_handler_for_dataframe(
+    df: Any,
+) -> TransformHandler[Any]:
+    """
+    Gets the handler for the given dataframe.
+
+    raises ValueError if the dataframe type is not supported.
+    """
+    if DependencyManager.has_pandas():
+        import pandas as pd
+
+        if isinstance(df, pd.DataFrame):
+            return PandasTransformHandler()
+    if DependencyManager.has_polars():
+        import polars as pl
+
+        if isinstance(df, pl.DataFrame):
+            return PolarsTransformHandler()
+
+    raise ValueError("Unsupported dataframe type. Must be Pandas or Polars.")
+
+
+class TransformsContainer(Generic[T]):
+    """
+    Keeps internal state of the last transformation applied to the dataframe.
+    So that we can incrementally apply transformations.
+    """
+
+    def __init__(self, df: T, handler: TransformHandler[T]) -> None:
+        self._original_df = df
+        # The dataframe for the given transform.
+        self._snapshot_df = df
+        self._handler = handler
+        self._transforms: List[Transform] = []
+
+    def apply(self, transform: Transformations) -> T:
+        """
+        Applies the given transformations to the dataframe.
+        """
+        # If the new transformations are a superset of the existing ones,
+        # then we can just apply the new ones to the snapshot dataframe.
+        if self._is_superset(transform):
+            transforms_to_apply = self._get_next_transformations(transform)
+            self._snapshot_df = apply_transforms(
+                self._snapshot_df, self._handler, transforms_to_apply
+            )
+            self._transforms = transform.transforms
+            return self._snapshot_df
+
+        # If the new transformations are not a superset of the existing ones,
+        # then we need to start from the original dataframe.
+        else:
+            self._snapshot_df = apply_transforms(
+                self._original_df, self._handler, transform
+            )
+            self._transforms = transform.transforms
+            return self._snapshot_df
+
+    def _is_superset(self, transforms: Transformations) -> bool:
+        """
+        Checks if the new transformations are a superset of the existing ones.
+        """
+        if not self._transforms:
+            return False
+
+        # If the new transformations are smaller than the existing ones,
+        # then it's not a superset.
+        if len(self._transforms) > len(transforms.transforms):
+            return False
+
+        for i, transform in enumerate(self._transforms):
+            if transform != transforms.transforms[i]:
+                return False
+
+        return True
```

Comment:
If you like one-liners:

```python
return all(t1 == t2 for t1, t1 in zip(self._transforms, transforms.transforms))
```

`zip` stops iteration at the length of the smaller iterable. Your code is fine as is though.